core/trino-main/src/main/java/io/trino/operator/aggregation/ApproximateSetGenericAggregation.java

```diff
@@ -0,0 +1,178 @@
+import static io.trino.spi.function.InvocationConvention.InvocationReturnConvention.FAIL_ON_NULL;
+import static io.trino.spi.function.OperatorType.XX_HASH_64;
+import static io.trino.util.Failures.internalError;
+
+@AggregationFunction(value = "approx_set_generic", hidden = true) // approx_set_generic is hidden as used only for statistics collection in ANALYZE so far; eventually it may be used for implementation of user visible approx_set
+public final class ApproximateSetGenericAggregation
+{
+    private static final int NUMBER_OF_BUCKETS = 4096;
+    private static final AccumulatorStateSerializer<HyperLogLogState> SERIALIZER = StateCompiler.generateStateSerializer(HyperLogLogState.class);
+
+    private ApproximateSetGenericAggregation() {}
+
+    @InputFunction
+    public static void input(
+            @AggregationState HyperLogLogState state,
+            @BlockPosition @SqlType("unknown") Block block,
+            @BlockIndex int index)
+    {
+        // do nothing -- unknown type is always NULL
+    }
+
+    @InputFunction
+    @TypeParameter("T")
+    public static void input(
+            @OperatorDependency(
+                    operator = XX_HASH_64,
+                    argumentTypes = "T",
+                    convention = @Convention(arguments = NEVER_NULL, result = FAIL_ON_NULL))
+                    MethodHandle methodHandle,
+            @AggregationState HyperLogLogState state,
+            @SqlType("T") double value)
+    {
+        HyperLogLog hll = getOrCreateHyperLogLog(state);
+        state.addMemoryUsage(-hll.estimatedInMemorySize());
+        long hash;
+        try {
+            hash = (long) methodHandle.invoke(value);
+        }
+        catch (Throwable t) {
+            throw internalError(t);
+        }
+        hll.addHash(hash);
+        state.addMemoryUsage(hll.estimatedInMemorySize());
+    }
+
+    @InputFunction
+    @TypeParameter("T")
+    public static void input(
+            @OperatorDependency(
+                    operator = XX_HASH_64,
+                    argumentTypes = "T",
+                    convention = @Convention(arguments = NEVER_NULL, result = FAIL_ON_NULL))
+                    MethodHandle methodHandle,
+            @AggregationState HyperLogLogState state,
+            @SqlType("T") long value)
+    {
+        HyperLogLog hll = getOrCreateHyperLogLog(state);
+        state.addMemoryUsage(-hll.estimatedInMemorySize());
+        long hash;
+        try {
+            hash = (long) methodHandle.invoke(value);
+        }
+        catch (Throwable t) {
+            throw internalError(t);
+        }
+        hll.addHash(hash);
+        state.addMemoryUsage(hll.estimatedInMemorySize());
+    }
+
+    @InputFunction
+    public static void input(@AggregationState HyperLogLogState state, @SqlType("BOOLEAN") boolean value)
+    {
+        HyperLogLog hll = getOrCreateHyperLogLog(state);
+        state.addMemoryUsage(-hll.estimatedInMemorySize());
+        hll.add(value ? 1 : 2);
```

Comment:
nit is `addHash` faster? :)

Reply:
Yeah - should be. We do not need to compute hash. Changed to use addHash with precomputed hash values.